pool/src/pool.rs

```diff
@@ -358,25 +353,27 @@ impl Pool {
 					// This is the common case for non 0-conf txs in the txpool.
 					// We assume the tx is valid here as we validated it on the way into the txpool.
 					insert_pos = Some(tx_buckets.len());
-					tx_buckets.push(entry.tx.clone());
+					tx_buckets.push(Bucket::new(entry.tx.clone()));
 				}
 				Some(pos) => {
 					// We found a single parent tx, so aggregate in the bucket
 					// if the aggregate tx is a valid tx.
 					// Otherwise discard and let the next block pick this tx up.
-					let current = tx_buckets[pos].clone();
-					if let Ok(agg_tx) = transaction::aggregate(vec![current, entry.tx.clone()]) {
-						if agg_tx
-							.validate(
-								Weighting::AsLimitedTransaction { max_weight },
-								self.verifier_cache.clone(),
-							)
-							.is_ok()
-						{
-							tx_buckets[pos] = agg_tx;
+					let bucket = &tx_buckets[pos];
+
+					if let Ok(new_bucket) = bucket.aggregate_with_tx(
+						entry.tx.clone(),
+						weighting,
+						self.verifier_cache.clone(),
+					) {
+						if new_bucket.fee_to_weight >= bucket.fee_to_weight {
+							// Only aggregate if it would not reduce the fee_to_weight ratio.
+							tx_buckets[pos] = new_bucket;
 						} else {
-							// Aggregated tx is not valid so discard this new tx.
-							is_rejected = true;
+							// Otherwise put it in its own bucket at the end.
+							// Note we increment the depth here to track the dependency.
+							tx_buckets
+								.push(Bucket::new_with_depth(entry.tx.clone(), bucket.depth + 1));
 						}
 					} else {
 						// Aggregation failed so discard this new tx.
```

Comment:
Out of curiosity out frequently this happens and how?

Reply:
The aggregation failure?

Basically the only way this would fail is if the aggregated tx ended up being too big to fit in a block. Say we attempted two huge transactions that in aggregate was larger than our block weight limit. This should happen only rarely.

The only other way this could happen is if the pool got itself into a bad state somehow with a double spend or something similar in there.
This should never happen in practice as we check for consistency at various stages of the tx pool lifecycle.